reco_utils/recommender/newsrec/models/naml.py

```diff
@@ -0,0 +1,272 @@
+            obj: a model used to evaluate and inference.
+        """
+
+        model, scorer = self._build_naml()
+        return model, scorer
+
+    def _build_userencoder(self, newsencoder):
+        '''The main function to create user encoder of NAML.
+
+        Args:
+            newsencoder(obj): the news encoder of NAML. 
+
+        Return:
+            obj: the user encoder of NAML.
+        '''
+        hparams = self.hparams
+        his_input_title_body_verts = keras.Input(shape=(hparams.his_size, 
+                                                hparams.title_size+hparams.body_size+2),dtype='int32')
+
+        click_news_presents = layers.TimeDistributed(newsencoder)(his_input_title_body_verts)
+        user_present = AttLayer2(hparams.attention_hidden_dim)(click_news_presents)
+        
+        model = keras.Model(his_input_title_body_verts, user_present, name='user_encoder')
+        return model
+    
+    def _build_newsencoder(self, embedding_layer):
+        '''The main function to create news encoder of NAML.
+        news encoder in composed of title encoder, body encoder, vert encoder and subvert encoder
+
+        Args:
+            embedding_layer(obj): a word embedding layer.
+        
+        Return:
+            obj: the news encoder of NAML.
+        '''
+        hparams = self.hparams
+        input_title_body_verts = keras.Input(shape=(hparams.title_size+hparams.body_size+2,),dtype='int32') 
+
+        sequences_input_title = layers.Lambda(lambda x: x[:, :hparams.title_size])(input_title_body_verts)
+        sequences_input_body = layers.Lambda(lambda x: x[:, hparams.title_size:hparams.title_size+hparams.body_size])(input_title_body_verts) 
+        input_vert = layers.Lambda(lambda x: x[:, hparams.title_size+hparams.body_size:hparams.title_size+hparams.body_size+1])(input_title_body_verts) 
+        input_subvert = layers.Lambda(lambda x: x[:, hparams.title_size+hparams.body_size+1:])(input_title_body_verts) 
+ 
+        title_repr = self._build_titleencoder(embedding_layer)(sequences_input_title)
+        body_repr = self._build_bodyencoder(embedding_layer)(sequences_input_body)
+        vert_repr = self._build_vertencoder()(input_vert)
+        subvert_repr = self._build_subvertencoder()(input_subvert)
+        
+        concate_repr = layers.Concatenate(axis=-2)([title_repr, body_repr, vert_repr, subvert_repr])
+        news_repr = AttLayer2(hparams.attention_hidden_dim)(concate_repr)
+        
+        model =  keras.Model(input_title_body_verts, news_repr, name="news_encoder")
+        return model
+    
+    def _build_titleencoder(self, embedding_layer):
+        '''build title encoder of NAML news encoder.
+
+        Args:
+            embedding_layer(obj): a word embedding layer.
+        
+        Return:
+            obj: the title encoder of NAML.
+        '''
+        hparams = self.hparams
+        sequences_input_title = keras.Input(shape=(hparams.title_size,),dtype='int32')   
```

Comment:
I noticed that now you are using keras instead of TF. Has the team moved to keras?

Reply:
Actually I use tf.keras. Different from Jianxun's team, the code of our(Fangzhao's) team is implemented by keras.

Reply:
got it thanks